### PR TITLE
feat(alert_policy_channels): add ability to add multiple channels to a policy

### PR DESCRIPTION
**Pull request by sanderblue**
Resolves: #319 

This PR introduces the ability to provision multiple alert channels per a policy via the `/v2/alerts_policy_channels` endpoint. Prior to this, we could only provision a single channel per HCL block. This will greatly reduce the amount of HCL code required to accomplish adding multiple notification channels to a policy.

### CODE REVIEW

File: newrelic/helpers.go [hunk 1]
```diff
@@ -27,6 +27,24 @@ func parseIDs(serializedID string, count int) ([]int, error) {
 	return ids, nil
 }
 
+// Converts a hash of IDs into an array.
+// Examples: "12345:54432:66564" -> []int{12345,54432,66564}
+func parseHashedIDs(serializedID string) ([]int, error) {
```
Comment (zlesnr) [inline newrelic/helpers.go]: Perhaps a test for this one would be good.

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 2]
```diff
@@ -136,3 +166,19 @@ func policyChannelExists(client *newrelic.NewRelic, policyID int, channelID int)
 
 	return false, nil
 }
+
+func policyChannelsExist(client *newrelic.NewRelic, policyID int, channelIDs []int) (bool, error) {
```
Comment (zlesnr) [inline newrelic/resource_newrelic_alert_policy_channel.go]: Perhaps here also a test would be good.

File: newrelic/structures_newrelic_alert_policy_channel.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package newrelic
+
+func expandChannelIDs(channelIDs []interface{}) []int {
```
Comment (zlesnr) [inline newrelic/structures_newrelic_alert_policy_channel.go]: A test for this one seems good too.

File: newrelic/helpers.go (hunk 1 above)
Comment (sanderblue) [reply newrelic/helpers.go]: Agreed

File: newrelic/structures_newrelic_alert_policy_channel.go (hunk 3 above)
Comment (sanderblue) [reply newrelic/structures_newrelic_alert_policy_channel.go]: Added

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 2 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: This is covered in integration tests 😎 Did you want a unit test for this?

Comment (zlesnr) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Oh not if its covered.

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 4]
```diff
@@ -83,35 +103,37 @@ func resourceNewRelicAlertPolicyChannelRead(d *schema.ResourceData, meta interfa
 	}
 
 	d.Set("policy_id", policyID)
-	d.Set("channel_id", channelID)
+	d.Set("channel_ids", channelIDs)
```
Comment (ctrombley) [inline newrelic/resource_newrelic_alert_policy_channel.go]: Do we need to set `channel_id` here?

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 5]
```diff
@@ -136,3 +158,19 @@ func policyChannelExists(client *newrelic.NewRelic, policyID int, channelID int)
 
 	return false, nil
 }
+
+func policyChannelsExist(client *newrelic.NewRelic, policyID int, channelIDs []int) (bool, error) {
+	for _, id := range channelIDs {
+		channelExists, err := policyChannelExists(client, policyID, id)
```
Comment (ctrombley) [inline newrelic/resource_newrelic_alert_policy_channel.go]: Can we batch this so we don't have to make a request each time for each iteration?

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 4 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: This is where the "interesting" logic of whether or not to set `channel_id` or `channel_ids` comes into play.

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 5 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Refactored to only make one request instead of multiple.

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 4 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Logic has been updated and is ready for re-review.

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 6]
```diff
@@ -83,56 +99,103 @@ func resourceNewRelicAlertPolicyChannelRead(d *schema.ResourceData, meta interfa
...
+		d.Set("channel_id", parsedChannelIDs[0])
+	}
+
+	if channelIDsOk && len(parsedChannelIDs) > 0 {
+		d.Set("channel_ids", parsedChannelIDs)
+	}
+
+	// Handle import (set `channel_ids` since this resource doesn't exist in state yet)
+	if !channelIDOk && !channelIDsOk {
+		d.Set("channel_ids", parsedChannelIDs)
```
Comment (zlesnr) [inline newrelic/resource_newrelic_alert_policy_channel.go]: Hmm.  In the case where a new resource has only `channel_id` set, does this tell them about an inaccurate create during the plan since we're setting the `channel_ids` plural?

File: newrelic/provider_test.go [hunk 7]
```diff
@@ -113,3 +114,32 @@ func testAccCreateApplication(t *testing.T) {
 
 // 	testAccCleanupComplete = true
 // }
+
+// testCheckResourceAttr validates the value in state for the given name/key combination.
+func testCheckResourceAttr(is *terraform.InstanceState, name string, key string, value string) error {
```
Comment (ctrombley) [inline newrelic/provider_test.go]: Why was this necessary to copy in?  [`TestCheckResourceAttr`](https://github.com/hashicorp/terraform/blob/master/helper/resource/testing.go#L955) is exported by the testing package, does it not get the job done in this case?

File: website/docs/r/alert_policy_channel.html.markdown [hunk 8]
```diff
@@ -38,4 +59,19 @@ resource "newrelic_alert_policy_channel" "foo" {
 The following arguments are supported:
 
   * `policy_id` - (Required) The ID of the policy.
-  * `channel_id` - (Required) The ID of the channel.
+  * `channel_ids` - (Optional*) Array of channel IDs to apply to the specified policy.
+  * `channel_id` - **Deprecated!** (Optional*) The ID of the channel. Please use the `channel_ids` argument instead.
+
+<sup>\*Note: Even though **channel_id** and **channel_ids** are optional, at least one of those arguments must be used for this resource to work.</sup>
+
+## Import
```
Comment (ctrombley) [inline website/docs/r/alert_policy_channel.html.markdown]: 👍

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 9]
```diff
@@ -24,55 +25,70 @@ func resourceNewRelicAlertPolicyChannel() *schema.Resource {
...
-	policyID := d.Get("policy_id").(int)
-	channelID := d.Get("channel_id").(int)
+	if err != nil {
+		return err
+	}
 
-	serializedID := serializeIDs([]int{policyID, channelID})
+	serializedID := serializeIDs(append(
+		[]int{policyChannels.ID},
+		policyChannels.ChannelIDs...,
```
Comment (ctrombley) [inline newrelic/resource_newrelic_alert_policy_channel.go]: If we don't sort these, changing their order in HCL will result in drift.   Probably not a big deal, but something to consider?

File: newrelic/resource_newrelic_alert_policy_channel_test.go [hunk 10]
```diff
@@ -32,6 +33,39 @@ func TestAccNewRelicAlertPolicyChannel_Basic(t *testing.T) {
 				),
 			},
 			// Test: Import
+			{
+				ResourceName:     resourceName,
+				ImportState:      true,
+				ImportStateCheck: testAccNewRelicAlertPolicyImportStateCheckFunc(),
```
Comment (ctrombley) [inline newrelic/resource_newrelic_alert_policy_channel_test.go]: Can this be replaced with a call to `TestCheckResourceAttr`?

File: newrelic/resource_newrelic_alert_policy_channel.go [hunk 11]
```diff
@@ -83,56 +99,103 @@ func resourceNewRelicAlertPolicyChannelRead(d *schema.ResourceData, meta interfa
 	}
 
 	d.Set("policy_id", policyID)
-	d.Set("channel_id", channelID)
+
+	_, channelIDOk := d.GetOk("channel_id")
```
Comment (ctrombley) [inline newrelic/resource_newrelic_alert_policy_channel.go]: Can some of the `Set`s here be refactored to a `flatten` func?

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 6 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Good call.

File: newrelic/resource_newrelic_alert_policy_channel_test.go (hunk 10 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel_test.go]: `ImportStateCheck` expects the type `ImportStateCheckFunc` which works with different types in the arguments that are passed to the pointer function.

File: newrelic/provider_test.go (hunk 7 above)
Comment (sanderblue) [reply newrelic/provider_test.go]: Well I could've written my own version of this, but this seemed like something we could use in other custom import check functions.

Comment (sanderblue) [reply newrelic/provider_test.go]: I can try to see if I can call one of the functions that TF exports inside the custom check I created. I tried several ways to do that, but couldn't get it to work like we'd want.

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 9 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Even if we sort the IDs and save them in order within the state, TF will still detect a change if the user changes the order of the IDs in the HCL in a way that doesn't match the state, which then creates a new resource due to `ForceNew: true`

Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: But sorting does avoid some form of drift, but there's still a chance a human change will cause drift. We can add a note in the docs to say "Best practice: sort your channel IDs in ascending order within your HCL to avoid state drift"

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 11 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: Updated

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 6 above)
Comment (sanderblue) [reply newrelic/resource_newrelic_alert_policy_channel.go]: After importing a subsequent `plan` will provide a deprecation warning message.

File: newrelic/provider_test.go (hunk 7 above)
Comment (sanderblue) [reply newrelic/provider_test.go]: Or I could do this
```go
func testAccNewRelicAlertPolicyImportStateCheckFunc(resourceName string, t *testing.T) resource.ImportStateCheckFunc {
	return func(state []*terraform.InstanceState) error {
		expectedChannelsCount := "1"
		channelsCount := state[0].Attributes["channel_ids.#"]

		if channelsCount != expectedChannelsCount {
			return fmt.Errorf(
				"%s: Attribute '%s' expected %#v, got %#v",
				resourceName,
				"channel_ids.#",
				expectedChannelsCount,
				channelsCount,
			)
		}

		return nil
	}
}
```

File: newrelic/resource_newrelic_alert_policy_channel.go (hunk 9 above)
Comment (ctrombley) [reply newrelic/resource_newrelic_alert_policy_channel.go]: 👍
